### ISSUE TITLE
Error when updating multiple properties in mapping of structs

### ISSUE THREAD

**Issue by giblfiz** [ethereum/solidity#724]
The following contract demonstrates this bug...
When funded the "works()" function behaves as expected,
but the "broken()" function... which differs only in that it updates a second value in the struct referenced by the mapping fails. 

`
contract micro {

```
struct R {
    uint  other;
    uint  count;
    uint  block_num;
}

mapping(address => R) public V;

function broken() {
    var twenty = 20;
    V[msg.sender].other += 19;
    V[msg.sender].count += twenty;
    V[msg.sender].block_num = block.number;
    msg.sender.send(twenty);
}

function works() {
    var twenty = 20;
    V[msg.sender].count += twenty;
    V[msg.sender].block_num = block.number;
    msg.sender.send(twenty);
}
```

}
`

**Comment by chriseth** [comment]
What does "work" and "fail" mean? What are your expectations towards those two functions and which kind of error do you get? Note that `works()` uses more gas that `broken()` - might that be the problem?

**Comment by giblfiz** [comment]
Thanks for the attention @chriseth 

"Works" updates the values of the components stored in the struct, and sends 20 wei to the message sender. (This can be observed by looking at balances, and by using the accessor method for V)

"Fail"... when broken is called... Does NOT update **any** of the elements stored in the struct, and does NOT send 20 wei. 

I have encountered nothing that would lead me to believe this is an "out of gas" problem... Mostly I have been testing calling from an account that has millions of wei on it, I generally put a a few hundred wei into the contract for testing as well. 

If you want I'm more than happy to link to you a copy of it on testnet.

Again, thanks for the help, this has been driving me absolutely nuts.

**Comment by chriseth** [comment]
Just tested this inside browser-solidity and could not find any oddness.

**Comment by giblfiz** [comment]
That's very odd... Here is a version of it I compiled on the online compiler (I'm using the one located here... http://ethereum.github.io/browser-solidity/#version=soljson-latest.js )

And pushed out to testnet...

`var m = eth.contract([{"constant":true,"inputs":[{"name":"","type":"address"}],"name":"V","outputs":[{"name":"other","type":"uint256"},{"name":"count","type":"uint256"},{"name":"block_num","type":"uint256"}],"type":"function"},{"constant":false,"inputs":[],"name":"broken","outputs":[],"type":"function"},{"constant":false,"inputs":[],"name":"works","outputs":[],"type":"function"}]
).at("0x11163a0988ed836762c00721e670a92d9580d7db")`

Which I have funded with 1000 Wei...

```
>eth.sendTransaction({from:eth.accounts[0], to:m.address, value:1000})

Tx(0x96791a59e886c56c25612bc0a09b12c502296e0d458893ba3ee9f54b690e5c0f) to: 0x11163a0988ed836762c00721e670a92d9580d7db
"0x96791a59e886c56c25612bc0a09b12c502296e0d458893ba3ee9f54b690e5c0f"

> eth.getBalance(m.address)
1000

> m.V.call(eth.accounts[0])
[0, 0, 0]


> m.broken.sendTransaction({from:eth.accounts[0]})
I0718 15:10:47.680345   38882 xeth.go:1028] 
Tx(0x9fee942a3ac2fc16f4caff1ac49ad938efde861e4f30abb12a0ba541394f7629) to: 0x11163a0988ed836762c00721e670a92d9580d7db
"0x9fee942a3ac2fc16f4caff1ac49ad938efde861e4f30abb12a0ba541394f7629

>"wait for blocks to mine"

> eth.getBalance(m.address)
1000

> m.V.call(eth.accounts[0])
[0, 0, 0]

> m.works.sendTransaction({from:eth.accounts[0]})
I0718 15:13:44.432772   38882 xeth.go:1028] Tx(0x02f659954d3ed191f9a161be7838933d6803f59a74ca22b3eec4c210edfaed26) to: 0x11163a0988ed836762c00721e670a92d9580d7db
"0x02f659954d3ed191f9a161be7838933d6803f59a74ca22b3eec4c210edfaed26"

>" wait for blocks to mine again" 

> eth.getBalance(m.address)
980

> m.V.call(eth.accounts[0])
[0, 20, 1332214]

```

Interestingly... if you call works() before you call broken() it seems like it works... 

Thanks for helping me out with with this @chriseth !!

**Comment by richbodo** [comment]
**Summary:**

I duplicated this particular bug.  

The function broken() does nothing when it uses the third element of the struct (the uint other).  Here is the address of a contract that I tested broken:

"0x0dcd2f752394c41875e259e00bb44fd505297caf"

The function broken() works as intended simply by removing the third element from the struct and not mentioning it in the function.  Here is the address of a contract that I tested working (third element removed):

"0xc8a70118e850b127ff76d4ba42e3e705f6d677fc"

The source of the contract that exhibits the broken behavior is the same as the contract @giblfiz posted called "micro", but I removed the working function and just used the broken() function.

The source of the contract that works just fine differs only by removing the uint "other" from the struct and line that uses the uint other from the function broken(), which is:

V[msg.sender].other += 19;

Those lines being as straightforward as they are, I can't explain why adding them breaks the function.

**Environment:**

Time: Duplicated a few minutes before this bug was filed.

Compiler: I'm using browser solidity from here: http://ethereum.github.io/browser-solidity/#version=soljson-latest.js

I'm using http instead of https because I can connect locally more easily that way.

On my local laptop, to fund and send transactions, I'm using the following go and geth:

Geth
Version: 1.3.6
Git Commit: 9e323d65b20c32ab1db40e8dde7172826e0e9c45
Protocol Versions: [63 62 61]
Network Id: 1
Go Version: go1.6.1
OS: darwin
GOPATH=/Users/rsb/src/gocode
GOROOT=/usr/local/Cellar/go/1.6.1/libexec

I'm on mac OSX 10.9.5
Browser is firefox: 47.0

Let me know if I can run more tests or be of any help.  Thanks!

**Comment by chriseth** [comment]
Could you try to reproduce the bug with a smaller contract (instead of trying to find workarounds ;-) ) or post explicit instructions on how to reproduce?

**Comment by giblfiz** [comment]
@chriseth 
Sadly, this seems to be the smallest contract that displays this particular bug. (You could remove the "working" function as it, well, works) 

The following, which was posted above, was intended to be an example of how to reproduce it. 
(Compiled with the "web-compiler", deployed on testnet, using the geth console to run commands)

```
>eth.sendTransaction({from:eth.accounts[0], to:m.address, value:1000})

Tx(0x96791a59e886c56c25612bc0a09b12c502296e0d458893ba3ee9f54b690e5c0f) to: 0x11163a0988ed836762c00721e670a92d9580d7db
"0x96791a59e886c56c25612bc0a09b12c502296e0d458893ba3ee9f54b690e5c0f"

> eth.getBalance(m.address)
1000

> m.V.call(eth.accounts[0])
[0, 0, 0]


> m.broken.sendTransaction({from:eth.accounts[0]})
I0718 15:10:47.680345   38882 xeth.go:1028] 
Tx(0x9fee942a3ac2fc16f4caff1ac49ad938efde861e4f30abb12a0ba541394f7629) to: 0x11163a0988ed836762c00721e670a92d9580d7db
"0x9fee942a3ac2fc16f4caff1ac49ad938efde861e4f30abb12a0ba541394f7629

>"wait for blocks to mine"

> eth.getBalance(m.address)
1000

> m.V.call(eth.accounts[0])
[0, 0, 0]

> m.works.sendTransaction({from:eth.accounts[0]})
I0718 15:13:44.432772   38882 xeth.go:1028] Tx(0x02f659954d3ed191f9a161be7838933d6803f59a74ca22b3eec4c210edfaed26) to: 0x11163a0988ed836762c00721e670a92d9580d7db
"0x02f659954d3ed191f9a161be7838933d6803f59a74ca22b3eec4c210edfaed26"

>" wait for blocks to mine again" 

> eth.getBalance(m.address)
980

> m.V.call(eth.accounts[0])
[0, 20, 1332214]
```

**Comment by chriseth** [comment]
@giblfiz this could very well be just an out-of-gas during the execution of "broken".  Do you have the tx receipt?

**Comment by giblfiz** [comment]
@chriseth ... OK I'm officially an idiot. 
Your right, it's an out of gas error ( https://testnet.etherscan.io/vmtrace?txhash=0x9fee942a3ac2fc16f4caff1ac49ad938efde861e4f30abb12a0ba541394f7629 )

I didn't realize (and no one mentioned/ it wasn't in the docs anywhere) that there is a default gas limit from an account. I hadn't even looked at the out-of-gas as a possibility since I could call "working / broken / working" again (I.E. plenty of gas in the sponsoring account) 

So in short:
to make this work I needed to call 
`bug.broken.sendTransaction({from:me, gas:100000})`

Thanks so much for your patience and your help!
